File: src/core/schema/tools/rename.rb

require 'core/system/library/schema'
require 'core/schema/tools/copy'

def paths(obj, path = [],  &block)
  field = ClassKey(obj.schema_class)

  current = path
  if field then
    key = obj[field.name]
    current = [*path, obj[field.name]]
    yield current
  end

  obj.schema_class.defined_fields.each do |f|
    next if f.computed
    next if !f.traversal
    if f.many then
      obj[f.name].each do |elt|
        paths(elt, current, &block) 
      end
    elsif f.optional then
      paths(obj[f.name], current, &block) unless obj[f.name].nil? || f.type.Primitive?
    else
      paths(obj[f.name], current, &block) unless f.type.Primitive?
    end
  end
end

def map_names!(obj, &block)
  field = ClassKey(obj.schema_class)

  if field then
    key = obj[field.name]
    x = yield obj[field.name]
    obj[field.name] = x if x
  end

  obj.schema_class.defined_fields.each do |f|
    next if f.computed
    next if !f.traversal
    if f.many then
      obj[f.name].each do |elt|
        map_names!(elt, &block) 
      end
    elsif f.optional then
      map_names!(obj[f.name], &block) unless obj[f.name].nil? || f.type.Primitive?
    else
      map_names!(obj[f.name], &block) unless f.type.Primitive?
    end
  end
end

def prime!(obj, suffix = '_')
  map_names!(obj) do |name|
    name + suffix
  end
end

def rename!(obj, map)
  field = ClassKey(obj.schema_class)

  # first recompute the nested rename map since
  # nested paths depend on the old names
  map2 = {}
  if field then
    map.each do |k, v|
      name = obj[field.name]
      if k.is_a?(Hash) && k[name] then
        map2[k[name]] = v
      else
        map2[k] = v
      end
    end
  else
    map2 = map
  end

  # then update the key of the current obj if there is a key
  # if it is in the map
  if field then
    key = obj[field.name]
    obj[field.name] = map[key] if map[key]
  end

  # then rename each non-primitive, non-nil, spine field using
  # the new map.
  obj.schema_class.defined_fields.each do |f|
    next if f.computed
    next if !f.traversal
    if f.many then
      obj[f.name].each do |elt|
        rename!(elt, map2) 
      end
    elsif f.optional then
      rename!(obj[f.name], map2) unless obj[f.name].nil? || f.type.Primitive?
    else
      rename!(obj[f.name], map2) unless f.type.Primitive?
    end
  end
end

def rename(obj, map)
  obj = copy(obj)
  rename!(obj, map)
  return obj
end


if __FILE__ == $0 then
  require 'core/system/load/load'
  require 'core/grammar/code/layout'
  require 'core/schema/code/factory'
  require 'core/schema/tools/print'
  require 'core/schema/tools/copy'
  ss = Loader.load('schema.schema')
  sg = Loader.load('schema.grammar')
  gg = Loader.load('grammar.grammar')
  gs = Loader.load('grammar.schema')

  ss_copy = copy(ss)

  map = {
    "Klass" => "Class",
    {"Klass" => "supers"} => "supertypes"
  }
  rename!(ss_copy, map)

  require 'core/grammar/code/rename_binding'
  
  sg_copy = copy(sg)
  rename_binding!(sg_copy, map)
  
  Print.print(sg_copy)

  DisplayFormat.print(gg, sg_copy)
  DisplayFormat.print(sg, ss_copy)

  prime!(sg_copy)
  
  puts "========= PRIMED ======="
  DisplayFormat.print(gg, sg_copy)

  paths(ss_copy) do |path|
    p path
  end
  
  #DisplayFormat.print(sg_copy, ss_copy)
end

File: src/js/core/schema/tools/rename.js
self.require("core/system/library/schema");
self.require("core/system/utils/paths");
self.require("core/schema/tools/copy");
(this.paths = (function (block, obj, path) {
  var self = this;
  (path = (((typeof path) !== "undefined") ? path : []));
  var current, field, key;
  (field = obj.schema_class().key());
  (current = path);
  if (field) {
    (key = obj._get(field.name()));
    (current = (path + obj._get(field.name())));
  }
  return obj.schema_class().defined_fields().each((function (f) {
    if (f.computed()) {
      self.next();
    }
    if ((!f.traversal())) {
      self.next();
    }
    if (f.many()) { 
      return obj._get(f.name()).each((function (elt) {
        return self.paths(block, elt, current);
      })); 
    }
    else { 
      if (f.optional()) { 
        if ((!((obj._get(f.name()) == null) || f.type().Primitive_P()))) {
          return self.paths(block, obj._get(f.name()), current);
        } 
      }
      else { 
        if ((!f.type().Primitive_P())) {
          return self.paths(block, obj._get(f.name()), current);
        }
      }
    }
  }));
}));
(this.map_names_C = (function (block, obj) {
  var self = this;
  var x, field, key;
  (field = obj.schema_class().key());
  if (field) {
    (key = obj._get(field.name()));
    (x = (function () {
    })());
    if (x) {
      obj._set(field.name(), x);
    }
  }
  return obj.schema_class().defined_fields().each((function (f) {
    if (f.computed()) {
      self.next();
    }
    if ((!f.traversal())) {
      self.next();
    }
    if (f.many()) { 
      return obj._get(f.name()).each((function (elt) {
        return self.map_names_C(block, elt);
      })); 
    }
    else { 
      if (f.optional()) { 
        if ((!((obj._get(f.name()) == null) || f.type().Primitive_P()))) {
          return self.map_names_C(block, obj._get(f.name()));
        } 
      }
      else { 
        if ((!f.type().Primitive_P())) {
          return self.map_names_C(block, obj._get(f.name()));
        }
      }
    }
  }));
}));
(this.prime_C = (function (obj, prefix) {
  var self = this;
  (prefix = (((typeof prefix) !== "undefined") ? prefix : "_"));
  return self.map_names_C((function (name) {
    return S("", prefix, ".", name, "");
  }), obj);
}));
(this.prime = (function (obj, prefix) {
  var self = this;
  (prefix = (((typeof prefix) !== "undefined") ? prefix : "_"));
  (obj = self.Clone(obj));
  self.prime_C(obj, prefix);
  return self.return(obj);
}));
(this.rename_C = (function (obj, map) {
  var self = this;
  var map2, field, key;
  (field = obj.schema_class().key());
  (map2 = (new EnsoHash({
    
  })));
  if (field) { 
    map.each((function (k, v) {
      var name;
      (name = obj._get(field.name()));
      if ((System.test_type(k, self.Hash()) && k._get(name))) { 
        return map2._set(k._get(name), v); 
      }
      else { 
        return map2._set(k, v);
      }
    })); 
  }
  else { 
    (map2 = map);
  }
  if (field) {
    (key = obj._get(field.name()));
    if (map._get(key)) {
      obj._set(field.name(), map._get(key));
    }
  }
  return self.rename_fields_C(obj, map2);
}));
(this.rename_fields_C = (function (obj, map) {
  var self = this;
  obj.schema_class().defined_fields().each((function (f) {
    if (f.computed()) {
      self.next();
    }
    if ((!f.traversal())) {
      self.next();
    }
    if (f.many()) {
      obj._get(f.name()).each((function (elt) {
        return self.rename_C(elt, map);
      }));
      if (f.type().key()) {
        return obj._get(f.name())._recompute_hash_C();
      }
    }
    else {
      if ((!((obj._get(f.name()) == null) || f.type().Primitive_P()))) {
        return self.rename_C(obj._get(f.name()), map);
      }
    }
  }));
  return obj;
}));
(this.rename = (function (obj, map) {
  var self = this;
  (obj = self.Clone(obj));
  self.rename_C(obj, map);
  return self.return(obj);
}));